design/one-pager-template-stacks.md

```diff
@@ -0,0 +1,679 @@
+
+The current design of Stacks allow for a new CRD to be managed by an independent
+controller. A CR will be handled by the Stack controller in any way that
+controller sees fit. This includes issuing remote API calls for additional
+inputs or creating external outputs, accessing local resources, or performing
+complex computations.
+
+Stacks may also opt to take advantage of the Crossplane Runtime to simplify the
+life cycle management of resources.
+
+In contrast, this proposal introduces Template Stacks with the limited ability
+to use their CR and child resources as input to create, update, or delete child
+Kubernetes resources. Application owners and managers can take advantage of
+Template Stacks to package complex deployments, while users benefit from a
+simple install and usage interface. Developers can use Template Stacks as
+application owners would, to create Kubernetes resource machinery using template
+engine logic, circumventing the need more complex controllers.
+
+The Template Stack Manager will permit the use of template variables to create
+and manage a set of resources provided as a set of strings included in the
+definition of the Stack.
+
+The benefits of Template Stacks are as follows:
+
+* [Simple Variable Substitution](#Simple-Variable-Substitution)
+
+  Stack Template resources can take input from a CR spec or from sibling
+  resources.
+* [No Custom Controller Needed](#No-Custom-Controller-Needed)
+
+  There is a single all-purpose controller that is designed to managed Template
+  Stacks. The existing Stack Manager will wire the Template Stack Manager for
+  managing the resources defined in a Template Stack.
+* [Composition](#Composition)
+
+  Stacks can be embedded in different Cluster environments, including other
+  Stacks. Charts and custom YAML files may reside in the cluster in the form of
+  a Template Stack.
+
+  Template Stacks introduce template variables that can take advantage of the
+  field values from other template rendered resources. Template variable
+  substitution is performed in each Kubernetes reconciliation pass, affected
+  all template resources for a given Stack CRD.
+
+### Simple Variable Substitution
+
+If an application owner wanted to parameterize a single `Deployment` within a
+more complex set of YAML resources, say the image version for a Redis container
+image, we could imagine placing a template variable in that `Deployment`.
+
+Supposing we used an inline handlebar template syntax, the `deployment` field
+could contain:
+
+```yaml
+containers:
+- name: redis-controller
+  image: redis/redis-crossplane-controller:{{.someVariable}}
+  imagePullPolicy: Always
+```
+
+Today, Kustomize overlays can be used to provide this level of functionality, as
+can Helm chart values. However, the capabilities provided by these tools are
+limited in two ways:
+
+* Original template values are not stored in the cluster
+* Replacement values can not be taken from other resources
```

Comment:
https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md 
is doing exactly this: taking values from other managed resources.

Reply:
I'll mention it here too.. Cross-resource referencing only works if the resource subscribes to the Crossplane development ideology, and only for predefined well known object types.

In Template Stacks, you can reuse arbitrary values from Crossplane and non-Crossplane resources.

I worry about expanding the scope of the benefits section to differentiate between all possible competitors, but I'll work this in here for now.

Reply:
Mentioned in 66983657